qoffer-infrastructure/src/main/groovy/life/qbic/portal/qoffer2/offers/OfferDbConnector.groovy

```diff
@@ -17,19 +22,95 @@ import org.apache.logging.log4j.Logger
  * @author: Jennifer Bödker
  *
  */
+@Log4j2
 class OfferDbConnector implements CreateOfferDataSource{
 
-    DatabaseSession session
+    ConnectionProvider connectionProvider
+
+    OfferToCustomerGateway offerToCustomerGateway
+    OfferToProductGateway offerToProductGateway
 
     private static final Logger LOG = LogManager.getLogger(OfferDbConnector.class)
 
-    OfferDbConnector(DatabaseSession session){
-        this.session = session
+    private static final String OFFER_INSERT_QUERY = "INSERT INTO offer (modificationDate, expirationDate, customerId, projectManagerId, projectTitle, projectDescription, totalPrice, customerAffiliationId)"
+
+
+    OfferDbConnector(ConnectionProvider connectionProvider, OfferToCustomerGateway offerToCustomerGateway, OfferToProductGateway offerToProductGateway){
+        this.connectionProvider = connectionProvider
+        this.offerToCustomerGateway = offerToCustomerGateway
+        this.offerToProductGateway = offerToProductGateway
     }
 
     @Override
     void store(Offer offer) throws DatabaseQueryException {
-        //TODO implement
-        throw new Exception("Method not implemented.")
+        try {
+            Connection connection = connectionProvider.connect()
+            connection.setAutoCommit(false)
+
+            connection.withCloseable {it ->
+                try {
+                    int projectManagerId = offerToCustomerGateway.getPersonId(offer.projectManager)
+                    int customerId = offerToCustomerGateway.getPersonId(offer.customer)
+                    int affiliationId = offerToCustomerGateway.getAffiliationId(offer.selectedCustomerAffiliation)
+
+                    int offerId = storeOffer(offer, projectManagerId, customerId, affiliationId)
+
+                    offerToProductGateway.createOfferItems(offer.items, offerId)
+                    connection.commit()
+                } catch (Exception e) {
+                    log.error(e.message)
+                    log.error(e.stackTrace.join("\n"))
+                    connection.rollback()
+                    connection.close()
+                    throw new DatabaseQueryException("Could not store offer.")
+                }
+            }
+        } catch (DatabaseQueryException ignored) {
+            throw new DatabaseQueryException("The offer could not be stored: ${offer.toString()}")
+        } catch (Exception e) {
+            log.error(e)
+            log.error(e.stackTrace.join("\n"))
+            throw new DatabaseQueryException("The offer could not be stored: ${offer.toString()}")
+        }
```

Comment:
three times the same error message thrown. I think this snippet can be much less verbose and less nested.

Reply:
dc8406c